# Python Pandas Training

##### This notebook will contain a basic guide on how to use pandas to manipulate and read in data from different data sources.
##### This notebook will be split into different sections depending on what we are doing with the data.

##### At the end of the notebook there will be some exercises (with the answers) to allow you to try out some of the functions mentioned for specific use cases.

In [1]:
# How to install modules in Jupyter Lab, make sure to restart the kernel afterwards.
# One liner install
%pip install numpy pandas xlsxwriter sqlalchemy pyodbc

# Installing over multiple lines
%pip install numpy
%pip install pandas
%pip install xlsxwriter
%pip install sqlalchemy
%pip install pyodbc

In [2]:
# Importing the pandas module.
# We use as to make it easier when writing code later.
import pandas as pd

## Read data into a datafarme.

##### There are multiple ways that we are able to read data in from different sources
##### The main 4 ways are shown below. When doing this with different data you might need to specify the delimiter or the unicode this can be found below or on the pandas website.

##### DataFrame: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html
##### CSV: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html
##### Excel: https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html
##### JSON: https://pandas.pydata.org/docs/reference/api/pandas.read_json.html
##### SQL: https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html

In [3]:
# Reading data from Python.
dictionary = {'First Column': ['String 1', 'String 2', 'String 3'],
            'Second Column': [1, 2, 3]}
df_dictionary = pd.DataFrame(data=dictionary)
df_dictionary

First Column  Second Column
0     String 1              1
1     String 2              2
2     String 3              3

In [4]:
# Reading data from numpy array.
import numpy as np

# Creating a numpy array.
np_array = np.array([[1, 2, 3],
                    [4, 5, 6],
                    [7, 8, 9]])

# Creating a dataframe with specific columns.
df_numpy = pd.DataFrame(np_array,
                        columns=['Column 1', 'Column 2', 'Column 3'])
df_numpy

Column 1  Column 2  Column 3
0         1         2         3
1         4         5         6
2         7         8         9

In [5]:
# Reading data from a csv file.
df_csv = pd.read_csv(r'Data/DoorDash CSV.csv', nrows=100)
df_csv

MONTH DELIVERY_STATE VECHICLE_TYPE          AGE       TIME_PERIOD  \
0   2017-01-01             AZ           car      26 - 35               UNK   
1   2017-01-01             AZ           car      26 - 35  a. 6 AM to 10 AM   
2   2017-01-01             AZ           car      26 - 35  b. 10 AM to 4 PM   
3   2017-01-01             AZ           car      36 - 50               UNK   
4   2017-01-01             AZ           car      36 - 50  a. 6 AM to 10 AM   
..         ...            ...           ...          ...               ...   
95  2017-01-01             MA           car  25 and Less  b. 10 AM to 4 PM   
96  2017-01-01             MA           car      26 - 35               UNK   
97  2017-01-01             MA           car      26 - 35   c. 4 PM to 8 PM   
98  2017-01-01             MA           car      36 - 50  a. 6 AM to 10 AM   
99  2017-01-01             MA           car      36 - 50  b. 10 AM to 4 PM   

   DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  D2R_DURATION  \
0      c. 250-1000               18+    Restaurant   2.0    196.200000   
1         d. 1000+               18+    Restaurant   2.0    436.833333   
2         d. 1000+      0 - 6 Months    Restaurant   2.0   1153.616667   
3        b. 50-250               18+    Restaurant   2.0    222.033333   
4        b. 50-250               18+    Restaurant   2.0    168.633333   
..             ...               ...           ...   ...           ...   
95            < 50               18+    Restaurant   3.0    198.050000   
96            < 50               18+    Restaurant   3.0    486.333333   
97     c. 250-1000               18+    Drive : Rx   3.0      0.000000   
98            < 50               18+    Restaurant   3.0     69.633333   
99        d. 1000+               18+    Drive : Rx   3.0      0.000000   

    R2C_DURATION  DASHER_WAIT_DURATION  R2C_DRIVEN_MILES  D2R_DRIVEN_MILES  \
0     277.600000            174.300000         62.241272         62.206983   
1     394.433333            295.483333        113.167706        161.458853   
2    1366.650000            510.650000        265.297382        204.501870   
3     342.416667            301.516667         75.299875         53.693890   
4     223.583333            128.500000         49.445137         39.240025   
..           ...                   ...               ...               ...   
95    145.416667             69.066667         18.803616         35.264339   
96    609.683333            315.816667         94.258728         96.557357   
97      0.000000              0.000000          9.437656          7.649002   
98    113.250000             71.100000         12.585411          6.174564   
99      0.000000              0.000000         54.096633         24.655237   

       GOV  DELIVERIES  
0    68611          17  
1   127915          41  
2   271861          86  
3    92997          21  
4    54904          18  
..     ...         ...  
95   69452          10  
96  170845          48  
97       0           1  
98   17379           7  
99       0          22  

[100 rows x 16 columns]

In [6]:
# Reading data from an excel file.
df_excel = pd.read_excel(r'Data/FuzzyMatching Excel.xlsx')
df_excel

No Model Found     Score  \
0                           032020Skeyes   1.000000   
1                       032020T2 Aviation  1.000000   
2                       032020T2 Aviation  1.000000   
3        052020Korea Aerospace Industries  1.000000   
4        052020Korea Aerospace Industries  1.000000   
..                                    ...       ...   
302  042020WTW / LS - Contingent Lineslip  0.402299   
303  042020WTW / LS - Contingent Lineslip  0.402299   
304  042020WTW / LS - Contingent Lineslip  0.402299   
305  042020WTW / LS - Contingent Lineslip  0.402299   
306         012020WLM / LS - Willis G-180  0.378788   

                           Best Match  Xuber Count  Pricing Model Count  \
0                        032020SKEYES            1                    1   
1                   032020T2 Aviation            2                    2   
2                   032020T2 Aviation            2                    2   
3    052020KOREA AEROSPACE INDUSTRIES            3                    1   
4    052020KOREA AEROSPACE INDUSTRIES            3                    1   
..                                ...          ...                  ...   
302         042021MONTENEGRO AIRLINES            4                    1   
303         042021MONTENEGRO AIRLINES            4                    1   
304         042021MONTENEGRO AIRLINES            4                    1   
305         042021MONTENEGRO AIRLINES            4                    1   
306           012020GALISTAIR TRADING            4                    1   

              Model of Best Match                Pricing Model Entries  \
0                    Aviation ATC                  092018TDG Aerospace   
1    Aviation Third Tier Airlines            012019YAKUTIA AIR COMPANY   
2    Aviation Third Tier Airlines                          010Tailwind   
3               Aviation Products  122018Empresa Nacional del Petroleo   
4               Aviation Products                          122018Meyer   
..                            ...                                  ...   
302  Aviation Third Tier Airlines          062020Thompson Aero Seating   
303  Aviation Third Tier Airlines                062020ISLAND AVIATION   
304  Aviation Third Tier Airlines    052020TRINIDAD & TOBAGO AIR GUARD   
305  Aviation Third Tier Airlines     072020Los Angeles World Airports   
306  Aviation Third Tier Airlines                    062020Air Niugini   

             Filtered No Model Found                   Matched Value  \
0                       032020skeyes                    032020skeyes   
1                   032020t2aviation                032020t2aviation   
2                   032020t2aviation                032020t2aviation   
3     052020koreaaerospaceindustries  052020koreaaerospaceindustries   
4     052020koreaaerospaceindustries  052020koreaaerospaceindustries   
..                               ...                             ...   
302  042020wtw/ls-contingentlineslip                       022020tag   
303  042020wtw/ls-contingentlineslip                       022020tag   
304  042020wtw/ls-contingentlineslip                       022020tag   
305  042020wtw/ls-contingentlineslip                       022020tag   
306         012020wlm/ls-willisg-180                       042020mda   

     Similarity Score  
0                 100  
1                 100  
2                 100  
3                 100  
4                 100  
..                ...  
302                60  
303                60  
304                60  
305                60  
306                60  

[307 rows x 10 columns]

In [7]:
# Reading data from a JSON file.
# When dealing with JSON files, make sure to be aware of what 'orientation' the data is in. 
df_json = pd.read_json(r'Data/AccountType JSON.json', orient='index')
df_json

0
aaa-gig                       Vehicle Leasing
dispatchit                           Delivery
doordash                             Delivery
doordashgl                           Delivery
fair                          Vehicle Leasing
fusionlogistics                    Amazon DSP
getaround                     Vehicle Leasing
gobrandsdbagopuff                    Delivery
grocerydeliverydbahellofresh         Delivery
grubhub                              Delivery
hyrecar                       Vehicle Leasing
maplebear_instacart                  Delivery
neighborfavor                        Delivery
neutronholdinglime              Micromobility
picomobility                    Micromobility
postmates                            Delivery
roadieinc                            Delivery
rydertruckrentalinc           Vehicle Leasing
shipt                                Delivery
unifiedfieldservices,inc.          Amazon DSP
walkercompaniesdbaprecise          Amazon DSP
wdcdelivery                          Delivery
weee!dbasaywee.com                   Delivery

#### Whenever we want to read data from SQL, we need to import sqlalchemy or create_engine from sqlalchemy.

In [8]:
# Importing sqlalchemy to read a database.
from sqlalchemy import create_engine

# Creating an engine to connect to.
engine = create_engine('mssql+pyodbc://azuprdsql013/APO_IBOTT?trusted_connection=yes&driver=ODBC+Driver+17+for+SQL+Server')

# Read data from a SQL database.
df_sql = pd.read_sql(f'''SELECT TOP (100) * FROM [APO_IBOTT].[dbo].[Claims_DoorDash]''', engine)
df_sql

ACCOUNT NAME ACCOUNT TYPE              ACCIDENT ID PREVIOUS ACCIDENT ID  \
0      DoorDash     Delivery   EMMANUEL MONTUNO_42795                 None   
1      DoorDash     Delivery         JADAN WING_42790                 None   
2      DoorDash     Delivery         JADAN WING_42790                 None   
3      DoorDash     Delivery         JADAN WING_42790                 None   
4      DoorDash     Delivery     LORENZO SHARIF_42804                 None   
..          ...          ...                      ...                  ...   
95     DoorDash     Delivery         JADAN WING_42790                 None   
96     DoorDash     Delivery         JADAN WING_42790                 None   
97     DoorDash     Delivery     LORENZO SHARIF_42804                 None   
98     DoorDash     Delivery  WILLIAM JEFFERSON_42763                 None   
99     DoorDash     Delivery       ALISHA CARIE_42790                 None   

   CLAIM ID VALUATION DATE POL EFF DATE POL EXP DATE        DOL TIME OF LOSS  \
0      None     2017-03-01         None         None 2017-03-01         None   
1      None     2017-03-01         None         None 2017-02-24         None   
2      None     2017-03-01         None         None 2017-02-24         None   
3      None     2017-03-01         None         None 2017-02-24         None   
4      None     2017-03-01         None         None 2017-03-10         None   
..      ...            ...          ...          ...        ...          ...   
95     None     2017-06-01         None         None 2017-02-24         None   
96     None     2017-06-01         None         None 2017-02-24         None   
97     None     2017-06-01         None         None 2017-03-10         None   
98     None     2017-06-01         None         None 2017-01-28         None   
99     None     2017-06-01         None         None 2017-02-24         None   

    ... VENDOR ORDER NUMBER  LITIGATION ACCOUNT MONTH MTD VEHICLE AGE  \
0   ...   None         None        None              None        None   
1   ...   None         None        None              None        None   
2   ...   None         None        None              None        None   
3   ...   None         None        None              None        None   
4   ...   None         None        None              None        None   
..  ...    ...          ...         ...               ...         ...   
95  ...   None         None        None              None        None   
96  ...   None         None        None              None        None   
97  ...   None         None        None              None        None   
98  ...   None         None        None              None        None   
99  ...   None         None        None              None        None   

   WRITTEN TIME AGING PENDING LOSS ADDRESS POLICY NUMBER   VIN  
0          None          None         None          None  None  
1          None          None         None          None  None  
2          None          None         None          None  None  
3          None          None         None          None  None  
4          None          None         None          None  None  
..          ...           ...          ...           ...   ...  
95         None          None         None          None  None  
96         None          None         None          None  None  
97         None          None         None          None  None  
98         None          None         None          None  None  
99         None          None         None          None  None  

[100 rows x 47 columns]

## Viewing Dataframe

##### There are multiple functions to get different information about a dataframe which will be shown below.
##### We'll be using the DoorDash delivery data for the examples but these functions can extend to any dataframe that you read in.

In [9]:
df = df_csv.copy()

In [10]:
# Head refers to the first rows in the dataframe.
df.head(3)

MONTH DELIVERY_STATE VECHICLE_TYPE      AGE       TIME_PERIOD  \
0  2017-01-01             AZ           car  26 - 35               UNK   
1  2017-01-01             AZ           car  26 - 35  a. 6 AM to 10 AM   
2  2017-01-01             AZ           car  26 - 35  b. 10 AM to 4 PM   

  DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  D2R_DURATION  \
0     c. 250-1000               18+    Restaurant   2.0    196.200000   
1        d. 1000+               18+    Restaurant   2.0    436.833333   
2        d. 1000+      0 - 6 Months    Restaurant   2.0   1153.616667   

   R2C_DURATION  DASHER_WAIT_DURATION  R2C_DRIVEN_MILES  D2R_DRIVEN_MILES  \
0    277.600000            174.300000         62.241272         62.206983   
1    394.433333            295.483333        113.167706        161.458853   
2   1366.650000            510.650000        265.297382        204.501870   

      GOV  DELIVERIES  
0   68611          17  
1  127915          41  
2  271861          86

In [11]:
# Tail refers to the last rows in the dataframe.
df.tail(3)

MONTH DELIVERY_STATE VECHICLE_TYPE      AGE       TIME_PERIOD  \
97  2017-01-01             MA           car  26 - 35   c. 4 PM to 8 PM   
98  2017-01-01             MA           car  36 - 50  a. 6 AM to 10 AM   
99  2017-01-01             MA           car  36 - 50  b. 10 AM to 4 PM   

   DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  D2R_DURATION  \
97     c. 250-1000               18+    Drive : Rx   3.0      0.000000   
98            < 50               18+    Restaurant   3.0     69.633333   
99        d. 1000+               18+    Drive : Rx   3.0      0.000000   

    R2C_DURATION  DASHER_WAIT_DURATION  R2C_DRIVEN_MILES  D2R_DRIVEN_MILES  \
97          0.00                   0.0          9.437656          7.649002   
98        113.25                  71.1         12.585411          6.174564   
99          0.00                   0.0         54.096633         24.655237   

      GOV  DELIVERIES  
97      0           1  
98  17379           7  
99      0          22

In [12]:
# Info gives essential details about your dataframe such as length, memory etc.
# When doing certain functions  with columns, you might need to check that the data in the columns is a specific data type.
# For example, making sure a number is an int/float type or datetime is in a datetime format.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   MONTH                 100 non-null    object 
 1   DELIVERY_STATE        100 non-null    object 
 2   VECHICLE_TYPE         100 non-null    object 
 3   AGE                   100 non-null    object 
 4   TIME_PERIOD           100 non-null    object 
 5   DELIVERY_TENURE       100 non-null    object 
 6   EXPERIENCE_TENURE     100 non-null    object 
 7   DELIVERY_TYPE         100 non-null    object 
 8   Tier                  100 non-null    float64
 9   D2R_DURATION          100 non-null    float64
 10  R2C_DURATION          100 non-null    float64
 11  DASHER_WAIT_DURATION  100 non-null    float64
 12  R2C_DRIVEN_MILES      100 non-null    float64
 13  D2R_DRIVEN_MILES      100 non-null    float64
 14  GOV                   100 non-null    int64  
 15  DELIVERIES            10

In [13]:
# Values gives an array of all of the values that exist in the "chosen" column.
df['TIME_PERIOD'].values

array(['UNK', 'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM', 'UNK',
       'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM', 'c. 4 PM to 8 PM',
       'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM', 'UNK', 'a. 6 AM to 10 AM',
       'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM', 'UNK', 'b. 10 AM to 4 PM',
       'UNK', 'UNK', 'c. 4 PM to 8 PM', 'a. 6 AM to 10 AM',
       'b. 10 AM to 4 PM', 'UNK', 'UNK', 'UNK', 'a. 6 AM to 10 AM',
       'a. 6 AM to 10 AM', 'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM',
       'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM',
       'c. 4 PM to 8 PM', 'c. 4 PM to 8 PM', 'a. 6 AM to 10 AM',
       'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM', 'c. 4 PM to 8 PM', 'UNK',
       'a. 6 AM to 10 AM', 'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM',
       'b. 10 AM to 4 PM', 'c. 4 PM to 8 PM', 'c. 4 PM to 8 PM',
       'c. 4 PM to 8 PM', 'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM',
       'b. 10 AM to 4 PM', 'b. 10 AM to 4 PM', 'UNK', 'UNK',
       'c. 4 PM to 8 PM', 'c. 4 PM to 8 PM', 'UNK', 'UNK', 'UNK',


In [14]:
# Unique gives the values that are unique in a specified column.
df['TIME_PERIOD'].unique()

array(['UNK', 'a. 6 AM to 10 AM', 'b. 10 AM to 4 PM', 'c. 4 PM to 8 PM'],
      dtype=object)

In [15]:
# Shape gives a tuple of the (rows, columns).
df.shape

(100, 16)

In [16]:
# Index gives us what the index of the dataframe is.
df.index

RangeIndex(start=0, stop=100, step=1)

In [17]:
# Columns gives us a list of the colum headers in our dataframe.
df.columns

Index(['MONTH', 'DELIVERY_STATE', 'VECHICLE_TYPE', 'AGE', 'TIME_PERIOD',
       'DELIVERY_TENURE', 'EXPERIENCE_TENURE', 'DELIVERY_TYPE', 'Tier',
       'D2R_DURATION', 'R2C_DURATION', 'DASHER_WAIT_DURATION',
       'R2C_DRIVEN_MILES', 'D2R_DRIVEN_MILES', 'GOV', 'DELIVERIES'],
      dtype='object')

In [18]:
# Corr creates a correlation matrix betwee each of the numerical columns in the dataframe.
df.corr()

Tier  D2R_DURATION  R2C_DURATION  \
Tier                  1.000000      0.118486      0.119181   
D2R_DURATION          0.118486      1.000000      0.994455   
R2C_DURATION          0.119181      0.994455      1.000000   
DASHER_WAIT_DURATION  0.118914      0.991873      0.998763   
R2C_DRIVEN_MILES      0.126533      0.995932      0.998561   
D2R_DRIVEN_MILES      0.121582      0.995333      0.996053   
GOV                   0.118310      0.996867      0.998654   
DELIVERIES            0.120016      0.996969      0.998864   

                      DASHER_WAIT_DURATION  R2C_DRIVEN_MILES  \
Tier                              0.118914          0.126533   
D2R_DURATION                      0.991873          0.995932   
R2C_DURATION                      0.998763          0.998561   
DASHER_WAIT_DURATION              1.000000          0.997891   
R2C_DRIVEN_MILES                  0.997891          1.000000   
D2R_DRIVEN_MILES                  0.996112          0.997969   
GOV                               0.996446          0.998312   
DELIVERIES                        0.998303          0.999197   

                      D2R_DRIVEN_MILES       GOV  DELIVERIES  
Tier                          0.121582  0.118310    0.120016  
D2R_DURATION                  0.995333  0.996867    0.996969  
R2C_DURATION                  0.996053  0.998654    0.998864  
DASHER_WAIT_DURATION          0.996112  0.996446    0.998303  
R2C_DRIVEN_MILES              0.997969  0.998312    0.999197  
D2R_DRIVEN_MILES              1.000000  0.996674    0.998493  
GOV                           0.996674  1.000000    0.999092  
DELIVERIES                    0.998493  0.999092    1.000000

In [19]:
# Describe gives you a summary of all the "numerical" colums in the dataframe.
df.describe()

Tier   D2R_DURATION   R2C_DURATION  DASHER_WAIT_DURATION  \
count  100.000000     100.000000     100.000000            100.000000   
mean     2.620000    5519.162000    8875.086667           4521.830667   
std      0.487832   23364.740318   38792.077638          19624.822866   
min      2.000000       0.000000       0.000000              0.000000   
25%      2.000000      49.312500      58.416667             35.087500   
50%      3.000000     352.291667     421.491667            271.583333   
75%      3.000000    1342.829166    1721.666667            886.179167   
max      3.000000  215369.216667  353099.033333         175413.433333   

       R2C_DRIVEN_MILES  D2R_DRIVEN_MILES           GOV    DELIVERIES  
count        100.000000        100.000000  1.000000e+02    100.000000  
mean        1682.053136       1256.745150  2.164680e+06    545.240000  
std         7513.793475       5344.491142  9.733202e+06   2378.291022  
min            0.357232          0.000000  0.000000e+00      1.000000  
25%           13.343205         10.082918  1.412600e+04      6.000000  
50%           73.399937         54.044264  1.089965e+05     29.500000  
75%          312.783822        373.461347  4.415070e+05    111.750000  
max        68369.985661      48179.377182  8.973368e+07  21599.000000

In [20]:
# Value counts gives the total number of occurences for each value in the column.
df['DELIVERY_STATE'].value_counts()
# df['VEHICLE_TYPE'].value_counts()

CA    44
IL    18
MA    12
AZ     8
DC     6
GA     6
CO     5
IN     1
Name: DELIVERY_STATE, dtype: int64

In [21]:
# Checks for null values in the whole dataframe.
# .sum() gives us nulls for the columns.
# .sum().sum() gives us the total number of nulls in the whole dataframe.
df_sql.isnull()#.sum()#.sum()

ACCOUNT NAME  ACCOUNT TYPE  ACCIDENT ID  PREVIOUS ACCIDENT ID  CLAIM ID  \
0          False         False        False                  True      True   
1          False         False        False                  True      True   
2          False         False        False                  True      True   
3          False         False        False                  True      True   
4          False         False        False                  True      True   
..           ...           ...          ...                   ...       ...   
95         False         False        False                  True      True   
96         False         False        False                  True      True   
97         False         False        False                  True      True   
98         False         False        False                  True      True   
99         False         False        False                  True      True   

    VALUATION DATE  POL EFF DATE  POL EXP DATE    DOL  TIME OF LOSS  ...  \
0            False          True          True  False          True  ...   
1            False          True          True  False          True  ...   
2            False          True          True  False          True  ...   
3            False          True          True  False          True  ...   
4            False          True          True  False          True  ...   
..             ...           ...           ...    ...           ...  ...   
95           False          True          True  False          True  ...   
96           False          True          True  False          True  ...   
97           False          True          True  False          True  ...   
98           False          True          True  False          True  ...   
99           False          True          True  False          True  ...   

    VENDOR  ORDER NUMBER  LITIGATION  ACCOUNT MONTH MTD  VEHICLE AGE  \
0     True          True        True               True         True   
1     True          True        True               True         True   
2     True          True        True               True         True   
3     True          True        True               True         True   
4     True          True        True               True         True   
..     ...           ...         ...                ...          ...   
95    True          True        True               True         True   
96    True          True        True               True         True   
97    True          True        True               True         True   
98    True          True        True               True         True   
99    True          True        True               True         True   

    WRITTEN TIME  AGING PENDING  LOSS ADDRESS  POLICY NUMBER   VIN  
0           True           True          True           True  True  
1           True           True          True           True  True  
2           True           True          True           True  True  
3           True           True          True           True  True  
4           True           True          True           True  True  
..           ...            ...           ...            ...   ...  
95          True           True          True           True  True  
96          True           True          True           True  True  
97          True           True          True           True  True  
98          True           True          True           True  True  
99          True           True          True           True  True  

[100 rows x 47 columns]

## Dataframe Manipulation

##### Once the data has been read in and we have our dataframe, we can start manipulating and analysing our data.
##### It should be mentioned that this section is basically unlimited in what you can do with it but I will try to go over some of the functions I believe are the most useful and effective.
##### A lot of these functions have more features within them, the documentation can be found below.

##### Concat: https://pandas.pydata.org/docs/reference/api/pandas.concat.html
##### Renaming Columns: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html
##### Drop NA: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html
##### Fill NA: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html
##### Drop Duplicates: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html
##### Merge: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html

In [22]:
# using concat to join different dataframes together.

# Two dataframes that have the same column headers.
df_concat = pd.concat([df_csv.head(2), df_csv.head(2)])

# Two dataframes that have different column headers.
df_concat2 = pd.concat([df_csv.head(2), df_excel.head(2)])

# Concatenating multiple dataframes.
dataframe_list = [df_csv, df_excel, df_json, df_sql]
df_concat3 = pd.concat(dataframe_list)

In [23]:
# Identical column concatenation.
df_concat

MONTH DELIVERY_STATE VECHICLE_TYPE      AGE       TIME_PERIOD  \
0  2017-01-01             AZ           car  26 - 35               UNK   
1  2017-01-01             AZ           car  26 - 35  a. 6 AM to 10 AM   
0  2017-01-01             AZ           car  26 - 35               UNK   
1  2017-01-01             AZ           car  26 - 35  a. 6 AM to 10 AM   

  DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  D2R_DURATION  \
0     c. 250-1000               18+    Restaurant   2.0    196.200000   
1        d. 1000+               18+    Restaurant   2.0    436.833333   
0     c. 250-1000               18+    Restaurant   2.0    196.200000   
1        d. 1000+               18+    Restaurant   2.0    436.833333   

   R2C_DURATION  DASHER_WAIT_DURATION  R2C_DRIVEN_MILES  D2R_DRIVEN_MILES  \
0    277.600000            174.300000         62.241272         62.206983   
1    394.433333            295.483333        113.167706        161.458853   
0    277.600000            174.300000         62.241272         62.206983   
1    394.433333            295.483333        113.167706        161.458853   

      GOV  DELIVERIES  
0   68611          17  
1  127915          41  
0   68611          17  
1  127915          41

In [24]:
# Non identical columns concatenation.
df_concat2

MONTH DELIVERY_STATE VECHICLE_TYPE      AGE       TIME_PERIOD  \
0  2017-01-01             AZ           car  26 - 35               UNK   
1  2017-01-01             AZ           car  26 - 35  a. 6 AM to 10 AM   
0         NaN            NaN           NaN      NaN               NaN   
1         NaN            NaN           NaN      NaN               NaN   

  DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  D2R_DURATION  ...  \
0     c. 250-1000               18+    Restaurant   2.0    196.200000  ...   
1        d. 1000+               18+    Restaurant   2.0    436.833333  ...   
0             NaN               NaN           NaN   NaN           NaN  ...   
1             NaN               NaN           NaN   NaN           NaN  ...   

      No Model Found  Score         Best Match  Xuber Count  \
0                NaN    NaN                NaN          NaN   
1                NaN    NaN                NaN          NaN   
0      032020Skeyes     1.0       032020SKEYES          1.0   
1  032020T2 Aviation    1.0  032020T2 Aviation          2.0   

   Pricing Model Count           Model of Best Match  \
0                  NaN                           NaN   
1                  NaN                           NaN   
0                  1.0                  Aviation ATC   
1                  2.0  Aviation Third Tier Airlines   

       Pricing Model Entries  Filtered No Model Found     Matched Value  \
0                        NaN                      NaN               NaN   
1                        NaN                      NaN               NaN   
0        092018TDG Aerospace             032020skeyes      032020skeyes   
1  012019YAKUTIA AIR COMPANY         032020t2aviation  032020t2aviation   

   Similarity Score  
0               NaN  
1               NaN  
0             100.0  
1             100.0  

[4 rows x 26 columns]

In [25]:
# Renaming columns in dataframe.
df.rename(columns={'D2R_DURATION': 'Driver to Restaurant Duration'}, inplace=True)
df.head(1)

MONTH DELIVERY_STATE VECHICLE_TYPE      AGE TIME_PERIOD  \
0  2017-01-01             AZ           car  26 - 35         UNK   

  DELIVERY_TENURE EXPERIENCE_TENURE DELIVERY_TYPE  Tier  \
0     c. 250-1000               18+    Restaurant   2.0   

   Driver to Restaurant Duration  R2C_DURATION  DASHER_WAIT_DURATION  \
0                          196.2         277.6                 174.3   

   R2C_DRIVEN_MILES  D2R_DRIVEN_MILES    GOV  DELIVERIES  
0         62.241272         62.206983  68611          17

In [26]:
df.columns = [col.lower() for col in df]
df.head(1)

month delivery_state vechicle_type      age time_period  \
0  2017-01-01             AZ           car  26 - 35         UNK   

  delivery_tenure experience_tenure delivery_type  tier  \
0     c. 250-1000               18+    Restaurant   2.0   

   driver to restaurant duration  r2c_duration  dasher_wait_duration  \
0                          196.2         277.6                 174.3   

   r2c_driven_miles  d2r_driven_miles    gov  deliveries  
0         62.241272         62.206983  68611          17

In [27]:
df.columns = ['Month', 'Delivery State', 'Vehicle Type', 'Age', 'Time Period', 'Delivery Tenure', 'Experience Tenure', 'Delivery Type', 
            'Tier', 'D2R Duration', 'R2C Duration', 'Dasher Wait Duration', 'R2C Driven Miles', 'D2R Driven Miles', 'Gov', 'Deliveries']
df.head(1)

Month Delivery State Vehicle Type      Age Time Period  \
0  2017-01-01             AZ          car  26 - 35         UNK   

  Delivery Tenure Experience Tenure Delivery Type  Tier  D2R Duration  \
0     c. 250-1000               18+    Restaurant   2.0         196.2   

   R2C Duration  Dasher Wait Duration  R2C Driven Miles  D2R Driven Miles  \
0         277.6                 174.3         62.241272         62.206983   

     Gov  Deliveries  
0  68611          17

In [28]:
df_month_index = df.copy()
# Set_index allows selection of which column you want to set as the index.
df_month_index.set_index('Month', inplace=True)

df_int_index = df.copy()

In [29]:
df_month_index.head(3)

Delivery State Vehicle Type      Age       Time Period  \
Month                                                               
2017-01-01             AZ          car  26 - 35               UNK   
2017-01-01             AZ          car  26 - 35  a. 6 AM to 10 AM   
2017-01-01             AZ          car  26 - 35  b. 10 AM to 4 PM   

           Delivery Tenure Experience Tenure Delivery Type  Tier  \
Month                                                              
2017-01-01     c. 250-1000               18+    Restaurant   2.0   
2017-01-01        d. 1000+               18+    Restaurant   2.0   
2017-01-01        d. 1000+      0 - 6 Months    Restaurant   2.0   

            D2R Duration  R2C Duration  Dasher Wait Duration  \
Month                                                          
2017-01-01    196.200000    277.600000            174.300000   
2017-01-01    436.833333    394.433333            295.483333   
2017-01-01   1153.616667   1366.650000            510.650000   

            R2C Driven Miles  D2R Driven Miles     Gov  Deliveries  
Month                                                               
2017-01-01         62.241272         62.206983   68611          17  
2017-01-01        113.167706        161.458853  127915          41  
2017-01-01        265.297382        204.501870  271861          86

In [30]:
df_int_index.head(3)

Month Delivery State Vehicle Type      Age       Time Period  \
0  2017-01-01             AZ          car  26 - 35               UNK   
1  2017-01-01             AZ          car  26 - 35  a. 6 AM to 10 AM   
2  2017-01-01             AZ          car  26 - 35  b. 10 AM to 4 PM   

  Delivery Tenure Experience Tenure Delivery Type  Tier  D2R Duration  \
0     c. 250-1000               18+    Restaurant   2.0    196.200000   
1        d. 1000+               18+    Restaurant   2.0    436.833333   
2        d. 1000+      0 - 6 Months    Restaurant   2.0   1153.616667   

   R2C Duration  Dasher Wait Duration  R2C Driven Miles  D2R Driven Miles  \
0    277.600000            174.300000         62.241272         62.206983   
1    394.433333            295.483333        113.167706        161.458853   
2   1366.650000            510.650000        265.297382        204.501870   

      Gov  Deliveries  
0   68611          17  
1  127915          41  
2  271861          86

In [31]:
# iloc is basically unlimited in how you can filter your dataframe but I'll only show how to select the number of records you want.
# Select the first to the second record. It's NON-INCLUSIVE of the second number.
df_subset = df_int_index.iloc[1:3]
#df_subset = df_month_index.iloc[1:3]
df_subset

Month Delivery State Vehicle Type      Age       Time Period  \
1  2017-01-01             AZ          car  26 - 35  a. 6 AM to 10 AM   
2  2017-01-01             AZ          car  26 - 35  b. 10 AM to 4 PM   

  Delivery Tenure Experience Tenure Delivery Type  Tier  D2R Duration  \
1        d. 1000+               18+    Restaurant   2.0    436.833333   
2        d. 1000+      0 - 6 Months    Restaurant   2.0   1153.616667   

   R2C Duration  Dasher Wait Duration  R2C Driven Miles  D2R Driven Miles  \
1    394.433333            295.483333        113.167706        161.458853   
2   1366.650000            510.650000        265.297382        204.501870   

      Gov  Deliveries  
1  127915          41  
2  271861          86

In [32]:
# Choosing certain columns.
df_subset2 = df_int_index['Month']
# df_subset2 = df_int_index[['Month']]
# df_subset2 = df_int_index[['Month', 'Delivery State', 'Vehicle Type', 'Age']]
df_subset2

0     2017-01-01
1     2017-01-01
2     2017-01-01
3     2017-01-01
4     2017-01-01
         ...    
95    2017-01-01
96    2017-01-01
97    2017-01-01
98    2017-01-01
99    2017-01-01
Name: Month, Length: 100, dtype: object

In [33]:
# How to filter column based off different conditions.

# Column being greater than a specific value.
df_filter = df_int_index[df_int_index['Deliveries'] > 10]

# Column being a specific value.
# df_filter = df_int_index[df_int_index['Time Period'] == 'a. 6 AM to 10 AM']
# df_filter = df_int_index[df_int_index['Month'] == '2017-01-01']

# Column having a value AND another column having a value.
# df_filter = df_int_index[(df_int_index['Delivery State'] == 'AZ') & (df_int_index['Deliveries'] > 40)]

# Column condition OR another column condition.
# df_filter = df_int_index[(df_int_index['Delivery State'] == 'AZ') | (df_int_index['Deliveries'] > 40) | (df_int_index['Vehicle Type'] == 'car')]
# df_filter = df_int_index[((df_int_index['Delivery State'] == 'AZ') | (df_int_index['Deliveries'] > 40)) & (df_int_index['Vehicle Type'] == 'car')]

# Value being in a column.
# df_filter = df_int_index[df_int_index['Time Period'].isin(['UNK'])]
df_filter

Month Delivery State Vehicle Type      Age       Time Period  \
0   2017-01-01             AZ          car  26 - 35               UNK   
1   2017-01-01             AZ          car  26 - 35  a. 6 AM to 10 AM   
2   2017-01-01             AZ          car  26 - 35  b. 10 AM to 4 PM   
3   2017-01-01             AZ          car  36 - 50               UNK   
4   2017-01-01             AZ          car  36 - 50  a. 6 AM to 10 AM   
..         ...            ...          ...      ...               ...   
89  2017-01-01             MA      bicycle  26 - 35               UNK   
91  2017-01-01             MA      bicycle  26 - 35   c. 4 PM to 8 PM   
94  2017-01-01             MA      bicycle  36 - 50  b. 10 AM to 4 PM   
96  2017-01-01             MA          car  26 - 35               UNK   
99  2017-01-01             MA          car  36 - 50  b. 10 AM to 4 PM   

   Delivery Tenure Experience Tenure Delivery Type  Tier  D2R Duration  \
0      c. 250-1000               18+    Restaurant   2.0    196.200000   
1         d. 1000+               18+    Restaurant   2.0    436.833333   
2         d. 1000+      0 - 6 Months    Restaurant   2.0   1153.616667   
3        b. 50-250               18+    Restaurant   2.0    222.033333   
4        b. 50-250               18+    Restaurant   2.0    168.633333   
..             ...               ...           ...   ...           ...   
89       b. 50-250               18+    Restaurant   3.0    575.866667   
91     c. 250-1000               18+    Restaurant   3.0   1622.416667   
94     c. 250-1000               18+    Restaurant   3.0    236.400000   
96            < 50               18+    Restaurant   3.0    486.333333   
99        d. 1000+               18+    Drive : Rx   3.0      0.000000   

    R2C Duration  Dasher Wait Duration  R2C Driven Miles  D2R Driven Miles  \
0     277.600000            174.300000         62.241272         62.206983   
1     394.433333            295.483333        113.167706        161.458853   
2    1366.650000            510.650000        265.297382        204.501870   
3     342.416667            301.516667         75.299875         53.693890   
4     223.583333            128.500000         49.445137         39.240025   
..           ...                   ...               ...               ...   
89    752.933333            402.700000         56.399626         49.223192   
91   2233.283333            877.483333        147.494389        146.940773   
94    171.150000             52.300000         20.629052         24.675187   
96    609.683333            315.816667         94.258728         96.557357   
99      0.000000              0.000000         54.096633         24.655237   

       Gov  Deliveries  
0    68611          17  
1   127915          41  
2   271861          86  
3    92997          21  
4    54904          18  
..     ...         ...  
89  129651          38  
91  420478         106  
94   41850          14  
96  170845          48  
99       0          22  

[65 rows x 16 columns]

In [34]:
# Drops rows where there are null values.
df.dropna(axis=0, inplace=True)#, how='all')

# Drops columns where there are null values.
df.dropna(axis=1, inplace=True)#, #how='all')

In [35]:
# Creating a numpy array with nuoll values inl.
null_np_array = np.array([[np.nan, 2, 3],
                            [4, np.nan, 6],
                            [7, 8, np.nan]])

df_empty_numpy = pd.DataFrame(null_np_array,
                        columns=['Column 1', 'Column 2', 'Column 3'])

df_empty_numpy_copy = df_empty_numpy.copy()

# Fill the null values in column 1 with the mean of the column.
df_empty_numpy_copy['Column 1'].fillna(df_empty_numpy_copy['Column 1'].mean(), inplace=True)
# Fill the null values in column 2 with 0.
df_empty_numpy_copy['Column 2'].fillna(0, inplace=True)
# Fill the null values in column 3 with infinity.
df_empty_numpy_copy['Column 3'].fillna(df_empty_numpy_copy['Column 2'], inplace=True)
# df_empty_numpy_copy['Column 3'].fillna(method='ffill', inplace=True)

In [36]:
df_empty_numpy

Column 1  Column 2  Column 3
0       NaN       2.0       3.0
1       4.0       NaN       6.0
2       7.0       8.0       NaN

In [37]:
df_empty_numpy_copy

Column 1  Column 2  Column 3
0       5.5       2.0       3.0
1       4.0       0.0       6.0
2       7.0       8.0       8.0

In [38]:
# Removing duplicated rows from the dataframe.
# There are no duplicates in the dataframe so I'm creating some to show the example.
print(f'Original shape is {df.shape}')

duplicated_df = pd.concat([df, df])
print(f'Duplicated shape is {duplicated_df.shape}')

# Dropping the duplicates and we can see we have been left with the original format.
duplicated_df.drop_duplicates(inplace=True)#, subset=['column'])
print(f'Removed duplicate shape is {duplicated_df.shape}')

Original shape is (100, 16)
Duplicated shape is (200, 16)
Removed duplicate shape is (100, 16)


In [39]:
# Sort_values orders the dataframe depending on a column(s).
# Axis=0 means rearrange the rows and axis=1 means rearrange the columns.
df.sort_values('Deliveries')#, axis=1, ascending=False)
# df.sort_values(['Month', 'Deliveries'])

Month Delivery State Vehicle Type          Age       Time Period  \
68  2017-01-01             GA          car          UNK  b. 10 AM to 4 PM   
12  2017-01-01             CA      bicycle      26 - 35  b. 10 AM to 4 PM   
52  2017-01-01             CO          car  25 and Less               UNK   
63  2017-01-01             GA          car  25 and Less               UNK   
44  2017-01-01             CA          car          UNK  b. 10 AM to 4 PM   
..         ...            ...          ...          ...               ...   
42  2017-01-01             CA          car      51 - 80   c. 4 PM to 8 PM   
24  2017-01-01             CA          car      26 - 35  a. 6 AM to 10 AM   
19  2017-01-01             CA          car  25 and Less  b. 10 AM to 4 PM   
36  2017-01-01             CA          car      51 - 80               UNK   
29  2017-01-01             CA          car      26 - 35  b. 10 AM to 4 PM   

   Delivery Tenure Experience Tenure Delivery Type  Tier   D2R Duration  \
68            < 50               18+    Restaurant   3.0      10.266667   
12     c. 250-1000               18+    Drive : Rx   3.0      10.416667   
52            < 50               18+    Restaurant   2.0       6.700000   
63       b. 50-250               18+    Restaurant   3.0       9.883333   
44            < 50               18+    Drive : Rx   3.0       0.000000   
..             ...               ...           ...   ...            ...   
42       b. 50-250               18+    Restaurant   3.0   21696.966667   
24        d. 1000+               18+    Restaurant   3.0   43682.916667   
19       b. 50-250               18+    Restaurant   3.0   40930.666667   
36        d. 1000+               18+    Restaurant   3.0   69726.166667   
29     c. 250-1000               18+    Restaurant   3.0  215369.216667   

     R2C Duration  Dasher Wait Duration  R2C Driven Miles  D2R Driven Miles  \
68       0.033333              9.666667          0.357232          0.357855   
12      36.416667              2.183333          4.183292          2.576060   
52       8.700000              4.366667          1.518703          1.873441   
63       9.650000             11.500000          0.877182          3.836035   
44       0.000000              0.000000          3.762469         18.485037   
..            ...                   ...               ...               ...   
42   40745.133333          20129.633333       7268.407107       4881.917706   
24   49949.100000          26437.000000      10794.836035      10467.349751   
19   70269.366667          35411.666667      14284.923940       9950.463217   
36  136456.433333          76710.333333      26259.038653      18901.527431   
29  353099.033333         175413.433333      68369.985661      48179.377182   

         Gov  Deliveries  
68      3972           1  
12         0           1  
52      2105           1  
63      8207           1  
44         0           1  
..       ...         ...  
42  10302510        2313  
24  14361382        3880  
19  17490557        4443  
36  30725300        8234  
29  89733682       21599  

[100 rows x 16 columns]

In [40]:
# Creating a function and applying it to a column.
def delivery_banding(x):
    if x >= 50:
        return "High Deliveries"
    else:
        return "Low Deliveries"

# Apply allows you to use any function with a column.
df['Delivery Band'] = df['Deliveries'].apply(delivery_banding)

# Inline version of the same function.
df['Delivery Band 2'] = df['Deliveries'].apply(lambda x: 'High Deliveries' if x >= 50 else 'Low Deliveries')

In [41]:
# Converting date column to datetime.
df['Datetime Month'] = pd.to_datetime(df['Month'])

In [42]:
# Creating a pivot table.
# Index is what you want the column values to be and values are what you want the rows to be.
# Aggfunc allows you to choose what operation you want to do on the column.
delivery_state_pivot = df.pivot_table(index='Delivery State',
                                        values='Deliveries',
                                        aggfunc={'Deliveries': np.sum})

# You can pivot multiple values as well as choosing what operation for each.
multiple_pivot = df.pivot_table(index='Delivery State',
                                        values=['Deliveries', 'D2R Driven Miles'],
                                        aggfunc={'Deliveries': np.mean,
                                                'D2R Driven Miles': np.mean})

In [43]:
delivery_state_pivot

Deliveries
Delivery State            
AZ                    1298
CA                   47190
CO                     127
DC                     171
GA                     101
IL                    5336
IN                       2
MA                     299

In [44]:
multiple_pivot

D2R Driven Miles   Deliveries
Delivery State                               
AZ                    482.447865   162.250000
CA                   2466.291983  1072.500000
CO                     68.470698    25.400000
DC                    351.981193    28.500000
GA                     53.917913    16.833333
IL                    558.494389   296.444444
IN                      7.538030     2.000000
MA                     38.324969    24.916667

In [45]:
# Getting data from SQL.
df_merge = df_sql.copy()
# Removing all columns which are full of null values.
df_merge.dropna(how='all', axis=1, inplace=True)
# Dropping the duplicate accident IDs.
df_merge.drop_duplicates(subset=['ACCIDENT ID'], inplace=True)
# Resetting the index value so it's easy to see how many rows are in the dataframe.
df_merge.reset_index(inplace=True, drop=True)

# Splitting into 2 different dataframes for merging later on.
df_left = df_merge[['ACCIDENT ID', 'VALUATION DATE', 'DOL', 'STATUS', 'OPEN/CLOSED', 'LOSS TYPE', 'LOSS SUB TYPE']].copy()
df_right = df_merge[['ACCIDENT ID', 'INCURRED RESERVE', 'INCURRED PAID', 'TOTAL INCURRED']].copy()

In [46]:
df_left.head(5)

ACCIDENT ID VALUATION DATE        DOL STATUS  OPEN/CLOSED  \
0   EMMANUEL MONTUNO_42795     2017-03-01 2017-03-01   OPEN          1.0   
1         JADAN WING_42790     2017-03-01 2017-02-24   OPEN          1.0   
2     LORENZO SHARIF_42804     2017-03-01 2017-03-10   OPEN          1.0   
3  WILLIAM JEFFERSON_42763     2017-03-01 2017-01-28   OPEN          1.0   
4       ALISHA CARIE_42790     2017-03-01 2017-02-24   OPEN          1.0   

   LOSS TYPE   LOSS SUB TYPE  
0  Liability  Liability - BI  
1  Liability  Liability - BI  
2        APD      APD - Coll  
3  Liability  Liability - PD  
4        APD      APD - Coll

In [47]:
df_right.head(5)

ACCIDENT ID  INCURRED RESERVE  INCURRED PAID  TOTAL INCURRED
0   EMMANUEL MONTUNO_42795            5500.0            0.0          5500.0
1         JADAN WING_42790            5500.0            0.0          5500.0
2     LORENZO SHARIF_42804            5500.0            0.0          5500.0
3  WILLIAM JEFFERSON_42763            5500.0            0.0          5500.0
4       ALISHA CARIE_42790            5500.0            0.0          5500.0

In [48]:
# Merging the 2 dataframes together on a common column that appears in both.
df_merged = df_left.merge(df_right, on='ACCIDENT ID')#, how='left')
df_merged.head(5)

ACCIDENT ID VALUATION DATE        DOL STATUS  OPEN/CLOSED  \
0   EMMANUEL MONTUNO_42795     2017-03-01 2017-03-01   OPEN          1.0   
1         JADAN WING_42790     2017-03-01 2017-02-24   OPEN          1.0   
2     LORENZO SHARIF_42804     2017-03-01 2017-03-10   OPEN          1.0   
3  WILLIAM JEFFERSON_42763     2017-03-01 2017-01-28   OPEN          1.0   
4       ALISHA CARIE_42790     2017-03-01 2017-02-24   OPEN          1.0   

   LOSS TYPE   LOSS SUB TYPE  INCURRED RESERVE  INCURRED PAID  TOTAL INCURRED  
0  Liability  Liability - BI            5500.0            0.0          5500.0  
1  Liability  Liability - BI            5500.0            0.0          5500.0  
2        APD      APD - Coll            5500.0            0.0          5500.0  
3  Liability  Liability - PD            5500.0            0.0          5500.0  
4        APD      APD - Coll            5500.0            0.0          5500.0

In [49]:
df_left2 = df_left.rename(columns={'ACCIDENT ID': 'ID'})
df_right2 = df_right.rename(columns={'ACCIDENT ID': 'ACCIDENT CODE'})

In [50]:
df_left2.head(5)

ID VALUATION DATE        DOL STATUS  OPEN/CLOSED  \
0   EMMANUEL MONTUNO_42795     2017-03-01 2017-03-01   OPEN          1.0   
1         JADAN WING_42790     2017-03-01 2017-02-24   OPEN          1.0   
2     LORENZO SHARIF_42804     2017-03-01 2017-03-10   OPEN          1.0   
3  WILLIAM JEFFERSON_42763     2017-03-01 2017-01-28   OPEN          1.0   
4       ALISHA CARIE_42790     2017-03-01 2017-02-24   OPEN          1.0   

   LOSS TYPE   LOSS SUB TYPE  
0  Liability  Liability - BI  
1  Liability  Liability - BI  
2        APD      APD - Coll  
3  Liability  Liability - PD  
4        APD      APD - Coll

In [51]:
df_right2.head(5)

ACCIDENT CODE  INCURRED RESERVE  INCURRED PAID  TOTAL INCURRED
0   EMMANUEL MONTUNO_42795            5500.0            0.0          5500.0
1         JADAN WING_42790            5500.0            0.0          5500.0
2     LORENZO SHARIF_42804            5500.0            0.0          5500.0
3  WILLIAM JEFFERSON_42763            5500.0            0.0          5500.0
4       ALISHA CARIE_42790            5500.0            0.0          5500.0

In [52]:
# Merging the dataframes together based off 2 different columns.
df_merged2 = df_left2.merge(df_right2, left_on='ID', right_on='ACCIDENT CODE')
df_merged2.head(5)

ID VALUATION DATE        DOL STATUS  OPEN/CLOSED  \
0   EMMANUEL MONTUNO_42795     2017-03-01 2017-03-01   OPEN          1.0   
1         JADAN WING_42790     2017-03-01 2017-02-24   OPEN          1.0   
2     LORENZO SHARIF_42804     2017-03-01 2017-03-10   OPEN          1.0   
3  WILLIAM JEFFERSON_42763     2017-03-01 2017-01-28   OPEN          1.0   
4       ALISHA CARIE_42790     2017-03-01 2017-02-24   OPEN          1.0   

   LOSS TYPE   LOSS SUB TYPE            ACCIDENT CODE  INCURRED RESERVE  \
0  Liability  Liability - BI   EMMANUEL MONTUNO_42795            5500.0   
1  Liability  Liability - BI         JADAN WING_42790            5500.0   
2        APD      APD - Coll     LORENZO SHARIF_42804            5500.0   
3  Liability  Liability - PD  WILLIAM JEFFERSON_42763            5500.0   
4        APD      APD - Coll       ALISHA CARIE_42790            5500.0   

   INCURRED PAID  TOTAL INCURRED  
0            0.0          5500.0  
1            0.0          5500.0  
2            0.0          5500.0  
3            0.0          5500.0  
4            0.0          5500.0

## Exporting Data

##### When you have a dataframe that you want to export you can choose what form and where to.

##### CSV: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html
##### Excel: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_excel.html
##### JSON: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html
##### SQL: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_sql.html

In [53]:
# Export to CSV with optional choice of separator.
df.to_csv(r'Export Folder/Dataframe To CSV.csv')#, sep=',')

# Export to Excel.
df.to_excel(r'Export Folder/Dataframe To Excel.xlsx')#, index=False, sheet_name='Sheet1')

# Export to JSON.
df.to_json(r'Export Folder/Dataframe To JSON.json')

# Export multiple dataframes to different sheets in Excel.
with pd.ExcelWriter(r'Export Folder/Multiple Sheet Export.xlsx', engine='xlsxwriter') as writer:
    delivery_state_pivot.to_excel(writer, sheet_name='Delivery Pivot')
    multiple_pivot.to_excel(writer, sheet_name='Delivery Duration Pivot')

# Export to an SQL table (only data people have write access).
# engine = create_engine('mssql+pyodbc://azuprdsql013/APO_IBOTT?trusted_connection=yes&driver=ODBC+Driver+17+for+SQL+Server')
# df.to_sql(table_name, con=engine, #if_exists='append')

## Exercise Time :)

### Make sure to read all the comments in the following cells for instructions of what to do next.

##### Below are some exercises to try and get used to the functions and syntax on a test dataset.
##### All of the exercises will be based off the functions earlier in this notebook.

In [54]:
# Firstly, we need data.
# Write code to read in the first 175 lines of GetAround data located in the Data folder.

# Finish the code below.
# HINT: What type of file is the GetAround data stored as? (JSON? CSV? EXCEL?)
# HINT: Make sure to specify the location (folders included).
GetAround_df = pd.read_csv(r'Data/GetAround CSV.csv', nrows=175)

##### Show the first 6 rows of the dataframe. What is the "Trip ID" of the last row? (ANSWER: 5592639518146560)

In [55]:
# Write code to show the first 6 rows of the dataframe.
# HINT: What sits on both your shoulders?
GetAround_df.head(6)

Trip ID         Renter ID  \
0  6463596273205248  6589678314717184   
1  5418439476117504   100004738281185   
2  5594859512922112  4936303853043712   
3  5453152066600960  5366438696452096   
4  6232742317916160   100004173131186   
5  5592639518146560  4718635344330752   

  Trip Start At \n            Local\n           Month  \
0                                            2021-11    
1                                            2021-11    
2                                            2021-11    
3                                            2021-11    
4                                            2021-11    
5                                            2021-11    

  Trip Start At \n            Local\n           Date  \
0                                         2021-11-11   
1                                         2021-11-10   
2                                         2021-11-06   
3                                         2021-11-23   
4                                         2021-11-10   
5                                         2021-11-20   

  Trip Created At \n            Local\n           Time  \
0                                2021-11-10 16:21:15     
1                                2021-11-10 15:49:41     
2                                2021-11-06 09:28:38     
3                                2021-11-23 12:02:42     
4                                2021-11-10 16:35:11     
5                                2021-11-19 17:31:51     

  Trip Start At \n            Local\n           Time  Trip Use  \
0                                2021-11-11 14:00:00  CARSHARE   
1                                2021-11-10 16:00:00  CARSHARE   
2                                2021-11-06 09:30:00  CARSHARE   
3                                2021-11-23 12:15:00  CARSHARE   
4                                2021-11-10 17:30:00  CARSHARE   
5                                2021-11-20 11:00:00  CARSHARE   

  Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
0                                 No                              No   
1                                 No                              No   
2                                 No                              No   
3                                 No                              No   
4                                 No                              No   
5                                 No                              No   

             Car VIN  Car Model Year Car Model Make Car Model Line  \
0  5N1AT2MV2KC844342            2019         Nissan          Rogue   
1  JTDKN3DP6E3054251            2014         Toyota  Prius Plug-In   
2  5YFEPMAE9MP196096            2021         Toyota        Corolla   
3  JM1DKFD70J0314674            2018          Mazda           CX-3   
4  JTLZE4FE6EJ058154            2014          Scion             xB   
5  JTDKDTB32F1099834            2015         Toyota        Prius c   

  Market Market Abbreviation    Market City Market State Abbreviation  \
0                    NYC-TRI       Brooklyn                        NY   
1                         SD      San Diego                        CA   
2                         SD      San Diego                        CA   
3                     SF-BAY  San Francisco                        CA   
4                     SF-BAY  San Francisco                        CA   
5                     SF-BAY       Berkeley                        CA   

   Renter Age     Renter Age Tier  Trip Sum Trip Duration Hours  \
0          30  T05 [30, 35) Years                          5.00   
1          35  T06 [35, 40) Years                          4.50   
2          24  T03 [23, 25) Years                         14.25   
3          32  T05 [30, 35) Years                          3.75   
4          38  T06 [35, 40) Years                          6.25   
5          49  T08 [45, 50) Years                         27.50   

  Trip Sum Miles Traveled  
0                      96  
1                      17  

##### Show the last 7 rows of the dataframe. What is the "Trip ID" of the first row? (ANSWER: 5837340984868864)

In [56]:
# Write code to show the last 7 rows of the dataframe.
# HINT: Dogs wag it.
GetAround_df.tail(7)

Trip ID         Renter ID  \
168  5837340984868864   100003950102263   
169  6564964889657344  5779005724033024   
170  5905275092860928  5912394783391744   
171  5461354789339136  6357141336489984   
172  5713024968425472  5551034431045632   
173  4566552939986944  4858802456363008   
174  4848431744286720  4977119392563200   

    Trip Start At \n            Local\n           Month  \
168                                            2021-11    
169                                            2021-11    
170                                            2021-11    
171                                            2021-11    
172                                            2021-11    
173                                            2021-11    
174                                            2021-11    

    Trip Start At \n            Local\n           Date  \
168                                         2021-11-22   
169                                         2021-11-22   
170                                         2021-11-25   
171                                         2021-11-15   
172                                         2021-11-25   
173                                         2021-11-02   
174                                         2021-11-23   

    Trip Created At \n            Local\n           Time  \
168                                2021-11-21 18:44:43     
169                                2021-11-22 12:33:28     
170                                2021-11-25 06:05:25     
171                                2021-11-14 07:30:41     
172                                2021-11-17 15:18:52     
173                                2021-11-02 07:41:22     
174                                2021-11-22 15:38:36     

    Trip Start At \n            Local\n           Time  Trip Use  \
168                                2021-11-22 06:30:00  CARSHARE   
169                                2021-11-22 12:45:00  CARSHARE   
170                                2021-11-25 06:15:00  CARSHARE   
171                                2021-11-15 15:15:00  CARSHARE   
172                                2021-11-25 09:00:00  CARSHARE   
173                                2021-11-02 17:30:00  CARSHARE   
174                                2021-11-23 06:45:00  CARSHARE   

    Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
168                                 No                              No   
169                                 No                              No   
170                                 No                              No   
171                                 No                              No   
172                                 No                              No   
173                                 No                              No   
174                                 No                              No   

               Car VIN  Car Model Year Car Model Make Car Model Line  \
168  KNMAT2MV0JP532757            2018         Nissan          Rogue   
169  5NPDH4AE5CH113848            2012        Hyundai        Elantra   
170  1G11A5SAXDF245808            2013      Chevrolet         Malibu   
171  JTDKB20U597841088            2009         Toyota          Prius   
172  JTKKU4B48C1018774            2012          Scion             xD   
173  3N1AB7AP1GY313578            2016         Nissan         Sentra   
174  19XFC1F32KE204401            2019          Honda    Civic Sedan   

    Market Market Abbreviation Market City Market State Abbreviation  \
168                    NYC-TRI    Brooklyn                        NY   
169                  BOS-GREAT   Hyde Park                        MA   
170                  HFD-METRO   New Haven                        CT   
171                    CHILAND     Chicago                        IL   
172                  BOS-GREAT    Brighton                        MA   
173                   DC-METRO  Washington                        DC   
174                  BOS-GREAT 

##### What is the shape of the GetAround dataframe? (ANSWER: (175, 20))

In [57]:
# Write code to show the rows and columns of the dataframe.
# HINT: A geometric figure (triangles, circles, squares, hexagons).
GetAround_df.shape

(175, 20)

##### What is the mean, minimum and maximum value of Renter Age? (ANSWER: Mean = 33.005714, Min = 19, Max = 74)

In [58]:
# Write code to find the summary statistics of Renter Age.
# HINT: It rhymes with PRESCRIBE.
GetAround_df.describe()

Trip ID     Renter ID  Car Model Year  Renter Age  \
count  1.750000e+02  1.750000e+02      175.000000  175.000000   
mean   5.593286e+15  4.762761e+15     2016.211429   33.005714   
std    6.353210e+14  2.197114e+15        3.530798    9.825195   
min    4.504769e+15  1.000039e+14     2005.000000   19.000000   
25%    5.069546e+15  4.926254e+15     2014.000000   26.000000   
50%    5.593574e+15  5.532596e+15     2016.000000   30.000000   
75%    6.119230e+15  6.119827e+15     2019.000000   37.000000   
max    6.729398e+15  6.753639e+15     2022.000000   74.000000   

       Trip Sum Trip Duration Hours  
count                    175.000000  
mean                      24.188571  
std                       42.203779  
min                        0.250000  
25%                        4.500000  
50%                       10.000000  
75%                       25.125000  
max                      392.000000

##### What is the total count of the value in Trip Use? (ANSWER: CARSHARE: 167, RIDESHARE: 8)

In [59]:
# Write code to find the value counts of the "Trip Use" column.
# HINT: Remeber to specify the column as well.
GetAround_df['Trip Use'].value_counts()

CARSHARE     167
RIDESHARE      8
Name: Trip Use, dtype: int64

##### How many nulls exist in the GetAround dataframe? (ANSWER: 0)

In [60]:
# Write code to find the total number of nulls in the dataframe.
# HINT: You'll write the same word twice.
GetAround_df.isna().sum().sum()

0

##### Neel accidentally split up a dataframe because he wasn't paying attention when coding. Please join them back together.

<img src="Pictures/DataFrame.jpg" width="600" height="600">

In [61]:
df_piece_1 = GetAround_df[:82].copy()
df_piece_2 = GetAround_df[82:].copy()

# Write code to join the dataframes together.
# HINT: Remember to keep the dataframes in a list.
df_glued_together = pd.concat([df_piece_1, df_piece_2])

# If you've written the code correctly, when you run the code below, it should output "True".
df_glued_together.equals(GetAround_df)

True

##### Neel has deemed the "Market State Abbreviation" column unworthy of the name it was given. Please rename it to "State".

In [62]:
# Write code to rename the "Market State Abbreviation" to "State".
# HINT: The format is {column_to_be_named: new_column_name}
GetAround_renamed = GetAround_df.rename(columns={'Market State Abbreviation': 'State'})

# If you've written the code correctly, when you run the code below, it should output 'Good Job'.
if 'State' in GetAround_renamed.columns:
    print('Good Job')
else:
    print('Keep trying my friend, I believe in you <3')

Good Job


##### We want to select a subset of the dataframe. We only want from the 2nd to the 7th (we want to include the 7th) row.
##### What is the first and last record of the subset? (ANSWER: First = 5418439476117504, Last = 6282717819174912)

In [63]:
# Finish the code to select the 2nd to the 7th row.
# HINT: Coders start counting from 0 not 1.
# HINT: The number to the right of the colon is NOT INCLUDED
GetAround_df.iloc[1:7]

Trip ID         Renter ID  \
1  5418439476117504   100004738281185   
2  5594859512922112  4936303853043712   
3  5453152066600960  5366438696452096   
4  6232742317916160   100004173131186   
5  5592639518146560  4718635344330752   
6  6282717819174912  6429152891240448   

  Trip Start At \n            Local\n           Month  \
1                                            2021-11    
2                                            2021-11    
3                                            2021-11    
4                                            2021-11    
5                                            2021-11    
6                                            2021-11    

  Trip Start At \n            Local\n           Date  \
1                                         2021-11-10   
2                                         2021-11-06   
3                                         2021-11-23   
4                                         2021-11-10   
5                                         2021-11-20   
6                                         2021-11-19   

  Trip Created At \n            Local\n           Time  \
1                                2021-11-10 15:49:41     
2                                2021-11-06 09:28:38     
3                                2021-11-23 12:02:42     
4                                2021-11-10 16:35:11     
5                                2021-11-19 17:31:51     
6                                2021-11-19 13:35:46     

  Trip Start At \n            Local\n           Time  Trip Use  \
1                                2021-11-10 16:00:00  CARSHARE   
2                                2021-11-06 09:30:00  CARSHARE   
3                                2021-11-23 12:15:00  CARSHARE   
4                                2021-11-10 17:30:00  CARSHARE   
5                                2021-11-20 11:00:00  CARSHARE   
6                                2021-11-19 14:00:00  CARSHARE   

  Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
1                                 No                              No   
2                                 No                              No   
3                                 No                              No   
4                                 No                              No   
5                                 No                              No   
6                                 No                              No   

             Car VIN  Car Model Year Car Model Make Car Model Line  \
1  JTDKN3DP6E3054251            2014         Toyota  Prius Plug-In   
2  5YFEPMAE9MP196096            2021         Toyota        Corolla   
3  JM1DKFD70J0314674            2018          Mazda           CX-3   
4  JTLZE4FE6EJ058154            2014          Scion             xB   
5  JTDKDTB32F1099834            2015         Toyota        Prius c   
6  JTDBT4K3XA1367790            2010         Toyota          Yaris   

  Market Market Abbreviation    Market City Market State Abbreviation  \
1                         SD      San Diego                        CA   
2                         SD      San Diego                        CA   
3                     SF-BAY  San Francisco                        CA   
4                     SF-BAY  San Francisco                        CA   
5                     SF-BAY       Berkeley                        CA   
6                    CHILAND        Chicago                        IL   

   Renter Age     Renter Age Tier  Trip Sum Trip Duration Hours  \
1          35  T06 [35, 40) Years                          4.50   
2          24  T03 [23, 25) Years                         14.25   
3          32  T05 [30, 35) Years                          3.75   
4          38  T06 [35, 40) Years                          6.25   
5          49  T08 [45, 50) Years                         27.50   
6          45  T08 [45, 50) Years                         42.00   

  Trip Sum Miles Traveled  
1                      17  
2                      83  

##### Neel is too lazy to count the individual rows of large Trip Sum Trip Duration Hours. Please filter the dataframe for Trip Sum Trip Duration Hours that are greater than 200.
##### What are the number of rows for this filtered dataframe? (ANSWER: 2)

In [64]:
# Finish the code to filter the dataframe for Trip Sum Trip Duration Hours.
# HINT: You'll have the dataframe appearing twice.
GetAround_df[GetAround_df['Trip Sum Trip Duration Hours'] > 200]

Trip ID         Renter ID  \
36  6377496070455296  5950255398256640   
84  6614682491551744  5325192240168960   

   Trip Start At \n            Local\n           Month  \
36                                            2021-11    
84                                            2021-11    

   Trip Start At \n            Local\n           Date  \
36                                         2021-11-24   
84                                         2021-11-28   

   Trip Created At \n            Local\n           Time  \
36                                2021-11-23 15:54:14     
84                                2021-11-28 08:01:28     

   Trip Start At \n            Local\n           Time  Trip Use  \
36                                2021-11-24 10:00:00  CARSHARE   
84                                2021-11-28 08:15:00  CARSHARE   

   Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
36                                 No                              No   
84                                 No                              No   

              Car VIN  Car Model Year Car Model Make Car Model Line  \
36  3MYDLBYV9LY708120            2020         Toyota    Yaris Sedan   
84  5NPDH4AE9BH060473            2011        Hyundai        Elantra   

   Market Market Abbreviation Market City Market State Abbreviation  \
36                  HOU-METRO     Houston                        TX   
84                         SD   San Diego                        CA   

    Renter Age     Renter Age Tier  Trip Sum Trip Duration Hours  \
36          32  T05 [30, 35) Years                         214.0   
84          25  T04 [25, 30) Years                         392.0   

   Trip Sum Miles Traveled  
36                   1,233  
84                     234

##### Neel would drive a Smart car to make up for how dumb he is. Please filter the dataframe Car Model Make for Smart cars.
##### Print the unique values in the column first before filtering to make sure we've spelt the car the correctly.
##### What are the number of rows for this filtered dataframe? (ANSWER: 4)

In [65]:
# Write code to filter the dataframe for Smart cars.
# HINT: Searching a column is CASE SENSITIVE.
# HINT: Print the unique values in the column first to find the cars that exist.
# HINT: Then filter the dataframe.
print(GetAround_df['Car Model Make'].unique())

# Don't type anything into this print, it's used for formatting.
print()

# Finish the line below to filter the code.
GetAround_df[GetAround_df['Car Model Make'] == 'smart']

['Nissan' 'Toyota' 'Mazda' 'Scion' 'Honda' 'Kia' 'Cadillac' 'Ford' 'FIAT'
 'Chevrolet' 'Volkswagen' 'Mitsubishi' 'Hyundai' 'smart' 'Acura' 'Jeep'
 'Subaru' 'Dodge' 'GMC' 'Volvo' 'Audi' 'Chrysler']



Trip ID         Renter ID  \
37   6149258555555840  5246031505784832   
123  4927333179195392  6105573430001664   
126  6350581725003776  6533485835714560   
130  4626620110733312   100007145371186   

    Trip Start At \n            Local\n           Month  \
37                                             2021-11    
123                                            2021-11    
126                                            2021-11    
130                                            2021-11    

    Trip Start At \n            Local\n           Date  \
37                                          2021-11-04   
123                                         2021-11-29   
126                                         2021-11-30   
130                                         2021-11-27   

    Trip Created At \n            Local\n           Time  \
37                                 2021-11-03 20:35:49     
123                                2021-11-29 14:13:39     
126                                2021-11-30 15:59:27     
130                                2021-11-27 17:16:38     

    Trip Start At \n            Local\n           Time  Trip Use  \
37                                 2021-11-04 08:45:00  CARSHARE   
123                                2021-11-29 15:30:00  CARSHARE   
126                                2021-11-30 16:30:00  CARSHARE   
130                                2021-11-27 18:45:00  CARSHARE   

    Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
37                                  No                              No   
123                                 No                              No   
126                                 No                              No   
130                                 No                              No   

               Car VIN  Car Model Year Car Model Make Car Model Line  \
37   WMEEJ31X48K206469            2008          smart         fortwo   
123  WMEEJ3BA8CK569795            2012          smart         fortwo   
126  WMEEJ3BA1FK806535            2015          smart         fortwo   
130  WMEEJ3BA7DK714097            2013          smart         fortwo   

    Market Market Abbreviation    Market City Market State Abbreviation  \
37                      SF-BAY  San Francisco                        CA   
123                     SF-BAY  San Francisco                        CA   
126                     SF-BAY  San Francisco                        CA   
130                     SF-BAY  San Francisco                        CA   

     Renter Age      Renter Age Tier  Trip Sum Trip Duration Hours  \
37           35   T06 [35, 40) Years                         16.00   
123          74  T12 [65, inf) Years                          2.50   
126          31   T05 [30, 35) Years                          1.75   
130          26   T04 [25, 30) Years                          4.50   

    Trip Sum Miles Traveled  
37                      103  
123                      14  
126                       8  
130                       0

##### Sort the dataframe by the Trip Sum Trip Duration Hours in descending order.
##### What is the largest Trip Sum Trip Duration Hours in the dataframe? (ANSWER: 392)

In [66]:
# Write code to sort the dataframe based of Trip Sum Trip Duration.
# HINT: Don't forget to include the "ascending" argument.
GetAround_df.sort_values('Trip Sum Trip Duration Hours', ascending=False)

Trip ID         Renter ID  \
84   6614682491551744  5325192240168960   
36   6377496070455296  5950255398256640   
163  6621567854706688  5889455639887872   
98   6110865306025984  6584881116479488   
110  5841795476226048  4692502632595456   
..                ...               ...   
43   4758842692206592   100006204651185   
91   5301813726543872  5898958402617344   
171  5461354789339136  6357141336489984   
165  4801940944191488  5166694859669504   
120  4918244290199552  5821707958222848   

    Trip Start At \n            Local\n           Month  \
84                                             2021-11    
36                                             2021-11    
163                                            2021-11    
98                                             2021-11    
110                                            2021-11    
..                                                 ...    
43                                             2021-11    
91                                             2021-11    
171                                            2021-11    
165                                            2021-11    
120                                            2021-11    

    Trip Start At \n            Local\n           Date  \
84                                          2021-11-28   
36                                          2021-11-24   
163                                         2021-11-23   
98                                          2021-11-11   
110                                         2021-11-28   
..                                                 ...   
43                                          2021-11-23   
91                                          2021-11-20   
171                                         2021-11-15   
165                                         2021-11-27   
120                                         2021-11-19   

    Trip Created At \n            Local\n           Time  \
84                                 2021-11-28 08:01:28     
36                                 2021-11-23 15:54:14     
163                                2021-11-22 11:38:05     
98                                 2021-11-11 11:46:43     
110                                2021-11-27 15:36:35     
..                                                 ...     
43                                 2021-11-23 09:04:07     
91                                 2021-11-20 10:05:03     
171                                2021-11-14 07:30:41     
165                                2021-11-27 08:21:28     
120                                2021-11-19 07:01:51     

    Trip Start At \n            Local\n           Time  Trip Use  \
84                                 2021-11-28 08:15:00  CARSHARE   
36                                 2021-11-24 10:00:00  CARSHARE   
163                                2021-11-23 17:45:00  CARSHARE   
98                                 2021-11-11 12:00:00  CARSHARE   
110                                2021-11-28 18:00:00  CARSHARE   
..                                                 ...       ...   
43                                 2021-11-23 09:15:00  CARSHARE   
91                                 2021-11-20 10:15:00  CARSHARE   
171                                2021-11-15 15:15:00  CARSHARE   
165                                2021-11-27 08:30:00  CARSHARE   
120                                2021-11-19 07:30:00  CARSHARE   

    Car Is Conveyance Owned (Yes / No) Car Is Claims Luxury (Yes / No)  \
84                                  No                              No   
36                                  No                              No   
163                                 No                              No   
98                                  No                              No   
110                                 No                              No   
..                                 ...                             ...   
43                                 

##### Create a pivot table with the index being the State and the values being the average of the Trip Sum Trip Duration Hours column.
##### What is the the average Trip Sum Trip Duration Hours forMA?? (ANSWER: 26.031250)

<img src="Pictures/Pivot Table.jpg" width=600 height=600 />

In [67]:
# Write code to create a pivot table with the State being the index and average Trip Sum Trip Duration Hours being the values.
# HINT: We need "index", "values" and "aggfunc".
# HINT: We want the average/mean of Trip Sum Trip Duration Hours.
Average_Trip_Sum_Pivot = GetAround_renamed.pivot_table(index='State',
                                        values='Trip Sum Trip Duration Hours',
                                        aggfunc={'Trip Sum Trip Duration Hours': np.mean})
Average_Trip_Sum_Pivot

Trip Sum Trip Duration Hours
State                              
AZ                         5.500000
CA                        20.487179
CO                        20.500000
CT                         9.000000
DC                        11.166667
FL                        22.625000
GA                        26.625000
IL                        33.625000
MA                        26.031250
NJ                         6.625000
NV                       156.000000
NY                        19.140000
OR                        13.875000
PA                        34.550000
TX                        70.300000
WA                         5.750000

##### Neel drank a bit too much and split the dataframe in half again. This time he split them vertically instead of horizontally. 
##### He also decided it would be a good idea to rename Trip ID on the second dataframe.
##### Please merge them back together.

<img src="Pictures/DataFrames 2.png" width=600 height=400 />

In [68]:
# Splitting the data in half based off columns.
df_1 = GetAround_df[['Trip ID', 'Renter ID', 'Trip Start At \n            Local\n           Month', 'Trip Start At \n            Local\n           Date',
                    'Trip Created At \n            Local\n           Time', 'Trip Start At \n            Local\n           Time', 'Trip Use', 'Car Is Conveyance Owned (Yes / No)',
                    'Car Is Claims Luxury (Yes / No)', 'Car VIN', 'Car Model Year', 'Car Model Make', 'Car Model Line']].copy()
df_2 = GetAround_df[['Trip ID', 'Market Market Abbreviation', 'Market City', 'Market State Abbreviation', 'Renter Age', 'Renter Age Tier', 'Trip Sum Trip Duration Hours', 'Trip Sum Miles Traveled']].copy()
df_2.rename(columns={'Trip ID': 'Trip Identifier'}, inplace=True)

# Write code to merge them back together.
# HINT: There are 2 ways to do this, you can either rename the column back to the original or you can use a left and right key, the result differs slightly but a merge is completed.
df_remerged = df_1.merge(df_2, left_on='Trip ID', right_on='Trip Identifier')

# df_2.rename(columns={'Trip Identifier': 'Trip ID'}, inplace=True)
# df_remerged = df_1.merge(df_2, on='Trip ID')

if 'Trip Identifier' in df_remerged.columns:
    df_remerged.drop(columns=['Trip Identifier'], inplace=True)

if df_remerged.equals(GetAround_df):
    print('Good Job')
else:
    print('Keep trying my friend, I believe in you <3')

Good Job
